### ISSUE TITLE
CMake: Linker PDB install rule does not work when Visual Studio generators are used

### ISSUE THREAD

**Issue by wittmeie** [zeromq/libzmq#2989]
*Please use this template for reporting suspected bugs or requests for help.*

# Issue description
1. When using the CMake Visual Studio solution generators, CMake generates multi-configuration projects. Hence, the variable `CMAKE_BUILD_TYPE` is not set and the CMake install rule for the linker PDB (CMakeLists.txt:1123) fails. Even setting the `CMAKE_BUILD_TYPE` variable explicitly does not solve the problem, as the PDB file is stored at a different location for multi-configuration projects.
2. The linker PDB is installed into `CMAKE_INSTALL_LIBDIR`. I guess `CMAKE_INSTALL_BINDIR` would be better as VS debuggers look for the PDB next to the DLL.

# Environment

* libzmq version (commit hash if unreleased): c9437ab755fc4f6962751e10da15ab1607ec2c36
* OS: Windows, CMake Visual Studio Generators

# Minimal test code / Steps to reproduce the issue

1.  Create a build folder
2.  Create the VS solution (e.g. `cmake -G"Visual Studio 15 2017 Win64" path-to-source`
3.  Build and install (`cmake --build . --target INSTALL`)

# What's the actual result? (include assertion message & call stack if applicable)
1.  If `CMAKE_BUILD_TYPE` is not set (the default), the PDB is not installed
2.  If `CMAKE_BUILD_TYPE=Debug`, the install step creates the following error 
` CMake Error at cmake_install.cmake:207 (file):
    file INSTALL cannot find
    "path-to-build-folder/bin/libzmq-v141-mt-gd-4_2_4.pdb".`

# What's the expected result?
Correct installation of the PDB

# Solution
Currently, ZeroMQ requires at least CMake 2.8.12. If you would switch to CMake 3.1.3, we could use CMake generator expressions - more specifically `$<TARGET_PDB_FILE:tgt>` (https://cmake.org/cmake/help/v3.1/manual/cmake-generator-expressions.7.html).
This would simply the install rule significantly as shown in zeromq/libzmq#2988

**Comment by sigiesec** [comment]
Just a thought: were it possible to require a CMake 3.1.3 only under Windows/Visual C++, if the commands requiring the new version are contained in an ``if (MSVC)`` block?

**Comment by bluca** [comment]
Bear in mind we have users on Windows XP, so only if it is available all the way down

**Comment by wittmeie** [comment]
I guess that could work. I could also check whether there is a way to determine the used CMake generator but that would also result in an additional if-statement (if single-configuration generator, else ...). Also not great. 
Is there a CI test for Windows XP which could validate whether 3.1.3 is working properly under XP? The installer is available here: https://cmake.org/files/v3.1/

**Comment by bluca** [comment]
Don't think we have any tests for XP unfortunately, as I'm not aware of any free-for-oss CI that provide XP images.

Is there any way to fix this bug without a new version of CMake? Doesn't matter if it's more verbose, as long as it does the job

**Comment by wittmeie** [comment]
No idea yet - but multi-configuration projects are inherently *evil* in my opinion! ;-) Without CMake handling the absolute path of the PDB there is almost no way to do this properly. And I think that was also the reason why generator expressions were added in CMake 3.x.

I was experimenting with explicitly setting the PDB_OUTPUT_DIRECTORY and the PDB_NAME properties but still, VS is adding subfolders below PDB_OUTPUT_DIRECTORY depending on the configuration (e.g. PDB_OUTPUT_DIRECTORY/Debug). I tested this with VS2017 but I am not even sure whether the subfolders are identical for older studio versions - and they are missing completely when single-configuration generators are used.

To cut a long story short - I do not have a solution.
